pybricks/tools/pb_module_tools.c

```diff
@@ -143,10 +144,88 @@ STATIC mp_obj_t pb_module_tools_run_task(size_t n_args, const mp_obj_t *pos_args
 }
 STATIC MP_DEFINE_CONST_FUN_OBJ_KW(pb_module_tools_run_task_obj, 1, pb_module_tools_run_task);
 
+#if PYBRICKS_PY_COMMON && PYBRICKS_PY_COMMON_LIGHT_MATRIX
+
+#include <pbsys/light.h>
+
+STATIC void pb_module_tools_hub_menu_display_symbol(mp_obj_t symbol) {
+    if (mp_obj_is_str(symbol)) {
+        pb_type_LightMatrix_display_char(pbsys_hub_light_matrix, symbol);
+    } else {
+        pb_type_LightMatrix_display_number(pbsys_hub_light_matrix, symbol);
+    }
+}
+
+STATIC pbio_button_flags_t pb_module_tools_hub_menu_wait_for_press(bool press) {
+    pbio_error_t err;
+    pbio_button_flags_t btn;
+    while ((err = pbio_button_is_pressed(&btn)) == PBIO_SUCCESS && ((bool)btn) == !press) {
+        MICROPY_EVENT_POLL_HOOK;
+    }
+    pb_assert(err);
+    return btn;
+}
+
+/**
+ * Displays a menu on the hub display and allows the user to pick a symbol
+ * using the buttons.
+ *
+ * @param [in]  n_args  The number of args.
+ * @param [in]  args    The args passed in Python code (the menu entries).
+ */
+STATIC mp_obj_t pb_module_tools_hub_menu(size_t n_args, const mp_obj_t *args) {
+
+    // Validate arguments by displaying all of them, ending with the first.
+    // This ensures we fail right away instead of midway through the menu. It
+    // happens so fast that there isn't a time penalty for this.
+    for (int i = n_args - 1; i >= 0; i--) {
+        pb_module_tools_hub_menu_display_symbol(args[i]);
+    }
+
+    // Disable stop button and cache original setting to restore later.
+    pbio_button_flags_t stop_button = pbsys_program_stop_get_buttons();
+    pbsys_program_stop_set_buttons(0);
```

Comment:
The rest of the code should probably be wrapped in `nlr_push()` (eventually to be replaced by https://github.com/micropython/micropython/pull/11445) so that the stop button gets restored no matter what.